src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs

```diff
@@ -0,0 +1,302 @@
+                // or in the processes of closing the buffer.
+                if (_currentState != EditorState.ProjectFileUpdateScheduledFromClean
+                    && _currentState != EditorState.ProjectFileUpdateScheduledFromDirty)
+                {
+                    return;
+                }
+
+                _currentState = _currentState == EditorState.ProjectFileUpdateScheduledFromDirty ?
+                    EditorState.ProjectFileChangingFromDirty : EditorState.ProjectFileChangingFromClean;
+            }
+
+            // Set set the buffer to be unmodifiable to prevent any changes while we update the project
+            await SetBufferReadonlyAsync(true).ConfigureAwait(true);
+
+            // TODO: Handle dirty state
+
+            // Set the buffer to default state and set the current state to editor clean
+            _textBufferManager.ResetBuffer();
+            await _textBufferStateListener.ForceBufferStateCleanAsync().ConfigureAwait(true);
+            lock (_lock)
+            {
+                _currentState = EditorState.EditorClean;
+            }
+            await SetBufferReadonlyAsync(false).ConfigureAwait(true);
+        }
+
+        #endregion
+
+        /// <summary>
+        /// Causes the dirty changes to the editor to be saved to the disk.
+        /// </summary>
+        public async Task SaveProjectFileAsync()
+        {
+            lock (_lock)
+            {
+                // In order to save, the editor must be in the dirty state.
+                if (_currentState != EditorState.EditorDirty) return;
+                _currentState = EditorState.WritingProjectFile;
+            }
+
+            // While saving the file, we disallow edits to the project file for sync purposes.
+            await SetBufferReadonlyAsync(true).ConfigureAwait(false);
+            await _textBufferStateListener.SaveAsync().ConfigureAwait(false);
+
+            lock (_lock)
+            {
+                _currentState = EditorState.EditorClean;
+            }
+            await SetBufferReadonlyAsync(false).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Sets the current editor state to either dirty or clean.
+        /// </summary>
+        public void SetEditorDirty(bool isDirty)
+        {
+            lock (_lock)
+            {
+                // If the current state isn't clean or dirty, some other change is happening, and we should not touch the current state
+                if (_currentState != EditorState.EditorClean && _currentState != EditorState.EditorDirty) return;
+
+                _currentState = _currentState == EditorState.EditorClean ? EditorState.EditorDirty : EditorState.EditorClean;
+            }
+        }
+
+        private async Task SetBufferReadonlyAsync(bool readOnly)
+        {
+            await _threadingService.SwitchToUIThread();
+            _textBufferManager.SetStateFlags(readOnly ? (uint)BUFFERSTATEFLAGS.BSF_USER_READONLY : 0);
```

Comment:
Should we be reading the "current" state flags here and setting them, otherwise we're overwriting them here?